libvast/src/format/arrow.cpp

```diff
@@ -74,4 +75,109 @@ bool writer::layout(const std::shared_ptr<::arrow::Schema>& schema) {
   return false;
 }
 
+arrow_istream_wrapper::arrow_istream_wrapper(std::shared_ptr<std::istream> input)
+  : input_(std::move(input)), pos_(0) {
+  set_mode(::arrow::io::FileMode::READ);
+}
+
+::arrow::Status arrow_istream_wrapper::Close() {
+  // TODO: is there a way to actually close an istream?
+  return ::arrow::Status::OK();
+}
+
+bool arrow_istream_wrapper::closed() const {
+  return input_->eof();
+}
+
+::arrow::Result<int64_t> arrow_istream_wrapper::Tell() const {
+  return pos_;
+}
+
+::arrow::Result<int64_t>
+arrow_istream_wrapper::Read(int64_t nbytes, void* out) {
+  input_->read(reinterpret_cast<char*>(out), nbytes);
+  if (input_) {
+    pos_ += nbytes;
+    return nbytes;
+  }
+  return 0;
+}
+
+::arrow::Result<std::shared_ptr<::arrow::Buffer>>
+arrow_istream_wrapper::Read(int64_t nbytes) {
+  ARROW_ASSIGN_OR_RAISE(auto buffer, ::arrow::AllocateResizableBuffer(nbytes));
+  ARROW_ASSIGN_OR_RAISE(int64_t bytes_read,
+                        Read(nbytes, buffer->mutable_data()));
+  ARROW_RETURN_NOT_OK(buffer->Resize(bytes_read, false));
+  buffer->ZeroPadding();
+  return std::move(buffer);
+}
+
+reader::reader(const caf::settings& options, std::unique_ptr<std::istream> in)
+  : vast::format::reader(options),
+    input_(std::make_unique<arrow_istream_wrapper>(std::move(in))) {
+}
+
+caf::error
+reader::read_impl(size_t max_events, size_t max_slice_size, consumer& f) {
+  VAST_TRACE_SCOPE("{} {}", VAST_ARG(max_events), VAST_ARG(max_slice_size));
+  VAST_ASSERT(max_events > 0);
+  // TODO: we currently ignore `max_slize_size` because we're just passing
+  // through existing table slices / record batches from the producer system.
+  VAST_ASSERT(max_slice_size > 0);
+  auto reader = ::arrow::ipc::RecordBatchStreamReader::Open(input_.get());
+  if (!reader.ok()) {
+    return caf::make_error(ec::logic_error, fmt::format("error creating stream "
+                                                        "reader: '{}'",
+                                                        reader.status()));
+  }
+  size_t produced = 0;
+  while (produced < max_events) {
```

Comment:
As far as I can tell this loop is missing the timeout handling the other readers have. That's important in case the input stream stalls in some way (`vast.import.read-timeout`), often leading to the import process not shutting down properly, or in case the user sets a low batch timeout (`vast.import.batch-timeout`).

Speaking of the the batch timeout and batch size options, I suppose it is intentional that this reader ignores them? Should we warn if the user did set custom options for them? Because of re-batching that mostly shouldn't matter, but if you intend to implement this then there's a re-batching function in the rebuild command that you can re-use.

Reply:
The initial plan was to make a relatively simple import, irgnoring the batch sizes alltogether and stick with the granularity of the upstream VAST node.

As you mention, rebuild should take care of mismatches in table slice size with the target node config.

I implemented the timeout checks, but find it difficult to test.

Reply:
I agree that we should not take the batch timeout and the batch size into consideration here.

I don't think your implementation works correctly for the read timeout, though. The call to get the next batch from the reader appears to be blocking, so it by definition does not respect the timeout.

Reply:
I see your point - what I've written there doesn't make much sense. I revisited the csv reader as my go-to reference implementation, and I see two different concepts wrt timing:

1. `read_timeout_` (20ms default); this seems to be the "real deal" in terms of stalled input. 
First of all, I don't see a way to integrate such a concept into the arrow record batch reader without introducing a separate thread / actor, because it's regular blocking I/O. 
Second, I'm in doubt whether this is actually what we want here. The use case of keeping one VAST node "in sync" with another one by doing 
`vast export --continuous arrow <query> | vast import arrow`. In such a case, stalling on read seems to be normal behavior, and not an exception.
2. `batch_timeout_`: we may not want to be involved in this for now, just taking the input batch granularity, relying on rebuild to optimize the data async. However, I wonder why the csv implementation appears to stop processing when
```c++
last_batch_sent_ + batch_timeout_ < reader_clock::now()
```
Shouldn't it just `finis()` and prepare for the next batch?

Reply:
Also @tobim @lava, your input would be highly appreciated on this since we spent so much time back then to implement the read timeouts correctly.

Reply:
After looking at this earlier today in a call together we quickly agreed to write a follow-up story for now and not to respect the read timeout for now.

Reply:
We agreed to delay the correct implementation of read timeouts to a later stage, probably as soon as we face the consequences of skipping this here. It's a bit tricky to implement, and the feature can already be useful.